url79d4f7c7ed_schanq_Live_Algorithm_tgz-pJ8-algorithms_and_systems_AI_Evolutionary_GeneticAlgorithmJ8/src/algorithms_and_systems_AI_Evolutionary/FuzzySystem.java

```diff
@@ -0,0 +1,938 @@
+												   ,this.AND(this.AND(onBeatUnarticulation,ascendingGradient),this.NOT(offBeatAccent)))
+										   ,this.OR(this.AND(this.AND(offBeatGhost, ascendingGradient),this.NOT(onBeatAccent))
+												   ,this.AND(this.AND(offBeatUnarticulation,ascendingGradient),this.NOT(onBeatAccent)))
+								     )));
+		double Sporadic    = this.OR(
+								this.OR(this.AND(this.AND(onBeatGhost,flatOrAsMakesNoDifference),this.NOT(offBeatAccent))
+									   ,this.AND(this.AND(onBeatUnarticulation,flatOrAsMakesNoDifference),this.NOT(offBeatAccent)))
+							   ,this.OR(this.AND(this.AND(offBeatGhost, flatOrAsMakesNoDifference),this.NOT(onBeatAccent))
+									   ,this.AND(this.AND(offBeatUnarticulation,flatOrAsMakesNoDifference),this.NOT(onBeatAccent)))
+									 );
+		
+		double On_Beat_Accents = this.OR(this.AND(this.AND(offBeatUnarticulation,flatOrAsMakesNoDifference),onBeatAccent),
+									this.OR(this.AND(this.AND(offBeatUnarticulation,descendingGradient),onBeatAccent),
+										this.OR(this.AND(this.AND(offBeatGhost,flatOrAsMakesNoDifference),onBeatAccent),
+											this.OR(this.AND(this.AND(offBeatGhost,descendingGradient),onBeatAccent),
+															this.AND(onBeatAccent,flatOrAsMakesNoDifference)))));
+		
+		double Off_Beat_Accents = this.OR(this.AND(this.AND(onBeatUnarticulation,flatOrAsMakesNoDifference),offBeatAccent),
+									this.OR(this.AND(this.AND(onBeatUnarticulation,descendingGradient),offBeatAccent),
+											this.OR(this.AND(this.AND(onBeatGhost,flatOrAsMakesNoDifference),offBeatAccent),
+												this.OR(this.AND(this.AND(onBeatGhost,descendingGradient),offBeatAccent),
+																this.AND(offBeatAccent,flatOrAsMakesNoDifference)))));
+		
+		// ...Whew! Now Defuzzify!!
+		double [] u = {Decrescendo,Crescendo,Sporadic,On_Beat_Accents,Off_Beat_Accents};
+		double [] x = {4,8,12,16,20};
+		double crisp = 0;
+		
+		// First check to see if there is an obvious peak in the results of the rule
+		// evaluations,If so then use then use the maximum method else use the singleton method. 			 		
+		if(this.defuzz_Max(u, x) > -1){
+			crisp = this.defuzz_Max(u, x);
+		}
+		else{	
+			crisp = this.defuzz_Singleton(u, x);
+		}
+		
+		// As this is defuzzifing into classifications a series of if
+		// then conditionals outlines the ranges of each type
+		if(crisp >= 0 && crisp <= 4)
+			d = DynamicShape.DECRESCENDO;
+		else if(crisp > 4 && crisp <=8)
+			d = DynamicShape.CRESCENDO;
+		else if(crisp > 8 && crisp <= 12)
+			d = DynamicShape.SPORADIC;
+		else if(crisp > 12 && crisp <= 16)
+			d = DynamicShape.ON_BEAT_ACCENTS;
+		else if(crisp > 16 && crisp <= 20)
+			d = DynamicShape.OFF_BEAT_ACCENTS;
+		
+		return d;
+		
+	}
+	// Method to determine what Dynamic enum is most true		
+	// for the performers current playing state. The 			
+	// output to be fed into the fitness function of the 		
+	// Genetic Algorithm 										
+	private Dynamic determineDynamic(int [] vels){
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 19. By comparison, 99% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test.

We recommend that you simplify this method or break it into multiple methods. For example, consider extracting the code block on lines 623-666 into a separate method.